paymaster/README.md

```diff
@@ -0,0 +1,23 @@
+## Paymaster
+
+This is a barebone verifying paymaster server. It is meant to be used in conjuction with the `VerifyingPaymaster` [contract](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/samples/VerifyingPaymaster.sol) by ETH Infinitism.
+
+The flow for sponsoring a user's transaction as follows:
+1. Users send a JSON RPC request to the paymaster server with the corresponding UserOp.
+2. The paymaster server signs the hash of the UserOp if it decides to sponsor the transaction. The signature is encoded in the `paymasterAndData` field of the userOp.
+3. The paymaster server sends the updated UserOp back to the user.
+4. The user can now send the UserOp to the bundler and the transaction will be sponsored by the paymaster.
```

Comment:
Oh I see, there is no mempool involved!

This is good, below is just me spitballing on what how I thought this would be + how it could be:

Is there a way in which the paymaster could submit the transaction to the bundler directly? I guess the updated userop needs to be signed by the user right?

Before seeing this, my very naive flow assumption was:

1. Users submit userOp to mempool (functionally equivalent, as its a RPC call, just that here we don't put it in the mempool)
2. Batchers pick userOp, check if they can be sponsored (using the whole 4337 dance) and if so submit them.
3. The sponsorship check is on-chain, the paymaster contract basically always sponsors non-reverting tx (is it even possible to check this in 4337???) to whitelisted contracts. Could also involve rate limiting (though update usage would have to live in the sponsored tx, NOT in the validation logic which can't write state).

I think the ideal flow for what we're doing would be to:

1. Send to server that is both a  paymaster & bundler. (Optionally drop to mempool, no need for now.)
2. Sign userOp, then wrap in a tx that
    - Writes to the paymaster contract, authorizing userOp
    - Submits the userOp to the entrypoint contract (paymaster validation then checks the authorization we just wrote)
    
This would be an ideal one-request flow from the user's perspective.

Reply:
You mean users send the signed UserOp to paymaster, and the paymaster submit to the entrypoint contract directly?

Reply:
Yes exactly!